### PR TITLE
Refactor `get_contracts_map`

### LINKED ISSUE #1005: Refactor `get_contracts_map` to take `Metadata` and `PackageId`
I think this function should just take `Metadata` and `PackageId` so we don't leak abstraction like in forge main (see below). Also the description is incorrect - path is not a path to scarb package, it is a path to `{target_name}.starknet_artifacts.json` file

https://github.com/foundry-rs/starknet-foundry/blob/master/crates/forge/src/main.rs#L168-L175

_Originally posted by @piotmag769 in https://github.com/foundry-rs/starknet-foundry/pull/957#discussion_r1371573537_

### PR DESCRIPTION
**Pull request by drknzz**
Closes #1005

## Introduced changes

- change input of **get_contracts_map**:
  - `Utf8Path` path to *{target_name}.starknet_artifacts.json*  --->  `Metadata` and `PackageId`

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cast/Cargo.toml [hunk 1]
```diff
@@ -16,7 +16,7 @@ url.workspace = true
 rand.workspace = true
 project-root.workspace = true
 which.workspace = true
-scarb-metadata = "=1.4.2"
+scarb-metadata = { git = "https://github.com/software-mansion/scarb.git", rev = "58cc88e" }
```
Comment (THenry14) [inline crates/cast/Cargo.toml:19]: I'd prefer this included changes suggested in https://github.com/foundry-rs/starknet-foundry/issues/675 instead

Comment (MaksymilianDemitraszek) [reply crates/cast/Cargo.toml:19]: Why do we need to set it to commit instead of version?

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Please split version bump and refactor as a separate PRs 🙏 . Such PRs are harder to review, merge quickly and create messy history. If we for example detected a regression later, we wouldn't be able to find out quickly if it happend from version bump or we messed somenting up doing refactor.

File: crates/cast/Cargo.toml (hunk 1 above)
Comment (piotmag769) [reply crates/cast/Cargo.toml:19]: @MaksymilianDemitraszek we don't need to anymore, we needed it before with Scarb 0.7.0 bcs scarb-ui was not published on crates.io and is dependent on metadata. As @THenry14 says, #675 is related to that

Comment (drknzz) [reply crates/cast/Cargo.toml]: Done in #1019

**Review by piotmag769 — APPROVED** [review]

**Review by THenry14 — APPROVED** [review]

File: crates/scarb-artifacts/src/lib.rs [hunk 2]
```diff
@@ -84,25 +84,34 @@ pub fn try_get_starknet_artifacts_path(
...
+                .ok_or_else(|| anyhow!("Failed to get parent for path = {}", &contracts_path))?;
+            let artifacts = artifacts_for_package(&contracts_path)?;
+            let mut map = HashMap::new();
+            for contract in artifacts.contracts {
+                let name = contract.contract_name;
+                let sierra_path = base_path.join(contract.artifacts.sierra);
+                let casm_path = base_path.join(contract.artifacts.casm);
+                let sierra = fs::read_to_string(sierra_path)?;
+                let casm = fs::read_to_string(casm_path)?;
+                map.insert(name, StarknetContractArtifacts { sierra, casm });
```
Comment (MaksymilianDemitraszek) [inline crates/scarb-artifacts/src/lib.rs:109]: Please add method `StarknetContractArtifacts::from_scarb_contract_artifact(StarknetContract)` I think it is the only way we read it, we can abstract this away.

Please move the whole logic under this pattern to `load_contract_artifacts` function, right now you need to read the whole code to understand what it does and cognitive complexity is high.

**Review by MaksymilianDemitraszek — APPROVED** [review]
LGTM, Please address the requested changes before merging 🙏 🥺

**Pull request merged**